### ISSUE TITLE
replace modelon/fmilib by NTNU-IHB/FMI4cpp 

### ISSUE THREAD

**Issue by prudhomm** [feelpp/feelpp#1747]
fmilib is a pain to maintain, cmake files are very old, complex and do not follow standard cmake builds. 
Also fmilib provides a C interface while we want to have a C++ one. We have a basic c++ interface in feelpp.
I propose to replace fmilib by fmi4cpp which provides a fmi2 interface for c++ 

for more information about fmi4cpp
https://github.com/NTNU-IHB/FMI4cpp

note that they have also pythonfmu for fmu packaging python scripts. Actually they use a C++ lib to make the FMI and wrap it. It could be used to package feelpp code as fmu

@vincentchabannes @romainhild thoughts ?

**Comment by vincentchabannes** [comment]
agree, seems good!

**Comment by prudhomm** [comment]
here is the code to init fmi4cpp

```cpp
auto fmu = fmi2::fmu( Environment::expand( this->M_coupledBC.back().circuit() ) ).as_cs_fmu()->new_instance();
this->M_circuits.push_back( fmu );
```

here is the code to run the fmi in coupledmixedpoisson:
```cpp
for( auto const& bc : M_coupledBC )
{
    auto md = M_circuits[i]->get_model_description();
    fmi2ValueReference vr_buffer = md->get_variable_by_name( bc.buffer() ).as_real().value_reference;
    fmi2Real v_buffer = this->M_mup[nbIbc+2*i+1]->max();
    M_circuits[i]->write_real( vr_buffer, v_buffer ) );
    M_circuits[i]->step(this->currentTime() );
    M_circuits[i]->read_real( vr_buffer, v_buffer ) );
    *this->M_mup[nbIbc+2*i+1] = project( _space=this->M_Ch, _expr=cst(v_buffer) );
    M_bdfsBuffer[i]->setUnknown( 0, *this->M_mup[nbIbc+2*i+1] );
    ++i;
}
```

**Comment by prudhomm** [comment]
@romainhild @thomas-saigre @mszopos here is what I obtain with fmi4cpp using hdg_coupledpoisson toolbox
on the linear test case using mo2fmu to generate the fmu.
seems good so far

```
$ ./feelpp_toolbox_hdg_coupledpoisson --config-file cases/coupledpoisson/1-linear/linear.cfg --ts.time-final 0.25 --hdg.poisson.json-editions BoundaryConditions.flux.Coupling.buffer.circuit:/home/prudhomm/Devel/feelpp/build/default/toolboxes/hdg/cases/coupledpoisson/1-linear/test3d0d.fmu 
Reading /home/prudhomm/Devel/feelpp/build/default/toolboxes/hdg/cases/coupledpoisson/1-linear/linear.cfg...
[ Starting Feel++ ] application mixed-poisson-model version 0.1 date 2022-Jan-06
 . mixed-poisson-model files are stored in /home/prudhomm/feel/toolboxes/hdg/coupledpoisson/linear/np_1
 .. logfiles :/home/prudhomm/feel/toolboxes/hdg/coupledpoisson/linear/np_1/logs
[modelProperties] Loading Model Properties : "/home/prudhomm/Devel/feelpp/build/default/toolboxes/hdg/cases/coupledpoisson/1-linear/linear.json"
[Info] [Thu Jan  6 13:16:02 2022
] /home/prudhomm/Devel/feelpp/feelpp/contrib/fmi4cpp/src/fmi4cpp/fmi2/fmi2_library.cpp:166: Calling fmi2SetupExperiment with toleranceDefined=false, tolerance=0, startTime=0, stopTimeDefined=false, stop=0
+------------------------------------------------------------------------------------------------------------------------------------------+
| Toolbox : poisson                                                                                                                        |
+------------------------------------------------------------------------------------------------------------------------------------------+
| +--------------------------------------------------------------------------------------+                                                 |
| | Environment                                                                          |                                                 |
| +--------------------------------------------------------------------------------------+                                                 |
| |  prefix              : hdg.poisson                                                   |                                                 |
| |  keyword             : poisson                                                       |                                                 |
| |  root repository     : /home/prudhomm/feel/toolboxes/hdg/coupledpoisson/linear/np_1  |                                                 |
| |  number of processus : 1                                                             |                                                 |
| +--------------------------------------------------------------------------------------+                                                 |
| +---------------------------+                                                                                                            |
| | Physics                   |                                                                                                            |
| +---------------------------+                                                                                                            |
| | +-----------+-----------+ |                                                                                                            |
| | | time mode | Transient | |                                                                                                            |
| | +-----------+-----------+ |                                                                                                            |
| +---------------------------+                                                                                                            |
| +------------------------------------------------------------------------+                                                               |
| | Materials Properties                                                   |                                                               |
| +------------------------------------------------------------------------+                                                               |
| |  number of materials : 1                                               |                                                               |
| | +--------------------------------------------------------------------+ |                                                               |
| | | Material : omega                                                   | |                                                               |
| | +--------------------------------------------------------------------+ |                                                               |
| | | +------+------------+----------------------+--------+------------+ | |                                                               |
| | | | Name | Expression | Symbol               | Shape  | Components | | |                                                               |
| | | +======+============+======================+========+============+ | |                                                               |
| | | | cond | k          | materials_omega_cond | scalar |            | | |                                                               |
| | | +------+------------+----------------------+--------+------------+ | |                                                               |
| | +--------------------------------------------------------------------+ |                                                               |
| +------------------------------------------------------------------------+                                                               |
| +--------------------------------------------------------------------------------------------------------------------------------------+ |
| | Meshes                                                                                                                               | |
| +--------------------------------------------------------------------------------------------------------------------------------------+ |
| | +----------------------------------------------------------------------------------------------------------------------------------+ | |
| | | Mesh : poisson                                                                                                                   | | |
| | +----------------------------------------------------------------------------------------------------------------------------------+ | |
| | |  Discretization : /Mesh/poisson                                                                                                  | | |
| | |  filename       : /home/prudhomm/feel/toolboxes/hdg/coupledpoisson/linear/np_1/hdg.poisson.msh                                   | | |
| | | +------------------------------------------------------------------------------------------------------------------------------+ | | |
| | | | Import configuration                                                                                                         | | | |
| | | +------------------------------------------------------------------------------------------------------------------------------+ | | |
| | | |  geo-filename          : /home/prudhomm/Devel/feelpp/build/default/toolboxes/hdg/cases/coupledpoisson/1-linear/test3d0d.geo  | | | |
| | | |  hsize                 : 1.000000e-01                                                                                        | | | |
| | | |  generate-partitioning : 0                                                                                                   | | | |
| | | +------------------------------------------------------------------------------------------------------------------------------+ | | |
| | | +-------------------------------+                                                                                                | | |
| | | | Discretization                |                                                                                                | | |
| | | +-------------------------------+                                                                                                | | |
| | | |  dim         : 3              |                                                                                                | | |
| | | |  h_average   : 1.647718e-01   |                                                                                                | | |
| | | |  h_max       : 2.113516e-01   |                                                                                                | | |
| | | |  h_min       : 1.000000e-01   |                                                                                                | | |
| | | |  n_edges     : 12439          |                                                                                                | | |
| | | |  n_elements  : 9126           |                                                                                                | | |
| | | |  n_faces     : 19452          |                                                                                                | | |
| | | |  n_partition : 1              |                                                                                                | | |
| | | |  n_points    : 2114           |                                                                                                | | |
| | | |  order       : 1              |                                                                                                | | |
| | | |  real_dim    : 3              |                                                                                                | | |
| | | |  shape       : Simplex_3_1_3  |                                                                                                | | |
| | | +-------------------------------+                                                                                                | | |
| | +----------------------------------------------------------------------------------------------------------------------------------+ | |
| +--------------------------------------------------------------------------------------------------------------------------------------+ |
| +-------------------------------------+                                                                                                  |
| | Function Spaces                     |                                                                                                  |
| +-------------------------------------+                                                                                                  |
| | +--------------------------------+  |                                                                                                  |
| | | Potential                      |  |                                                                                                  |
| | +--------------------------------+  |                                                                                                  |
| | |  mesh   : /Mesh/poisson        |  |                                                                                                  |
| | |  nSpace : 1                    |  |                                                                                                  |
| | | +----------------------------+ |  |                                                                                                  |
| | | | Basis                      | |  |                                                                                                  |
| | | +----------------------------+ |  |                                                                                                  |
| | | |  is_continuous : 0         | |  |                                                                                                  |
| | | |  nComponents   : 1         | |  |                                                                                                  |
| | | |  nComponents1  : 1         | |  |                                                                                                  |
| | | |  nComponents2  : 1         | |  |                                                                                                  |
| | | |  nLocalDof     : 4         | |  |                                                                                                  |
| | | |  name          : lagrange  | |  |                                                                                                  |
| | | |  order         : 1         | |  |                                                                                                  |
| | | |  shape         : scalar    | |  |                                                                                                  |
| | | +----------------------------+ |  |                                                                                                  |
| | | +----------------+             |  |                                                                                                  |
| | | | Dof Table      |             |  |                                                                                                  |
| | | +----------------+             |  |                                                                                                  |
| | | |  nDof : 36504  |             |  |                                                                                                  |
| | | +----------------+             |  |                                                                                                  |
| | +--------------------------------+  |                                                                                                  |
| | +---------------------------------+ |                                                                                                  |
| | | Flux                            | |                                                                                                  |
| | +---------------------------------+ |                                                                                                  |
| | |  mesh   : /Mesh/poisson         | |                                                                                                  |
| | |  nSpace : 1                     | |                                                                                                  |
| | | +-----------------------------+ | |                                                                                                  |
| | | | Basis                       | | |                                                                                                  |
| | | +-----------------------------+ | |                                                                                                  |
| | | |  is_continuous : 0          | | |                                                                                                  |
| | | |  nComponents   : 3          | | |                                                                                                  |
| | | |  nComponents1  : 3          | | |                                                                                                  |
| | | |  nComponents2  : 1          | | |                                                                                                  |
| | | |  nLocalDof     : 4          | | |                                                                                                  |
| | | |  name          : lagrange   | | |                                                                                                  |
| | | |  order         : 1          | | |                                                                                                  |
| | | |  shape         : vectorial  | | |                                                                                                  |
| | | +-----------------------------+ | |                                                                                                  |
| | | +-----------------+             | |                                                                                                  |
| | | | Dof Table       |             | |                                                                                                  |
| | | +-----------------+             | |                                                                                                  |
| | | |  nDof : 109512  |             | |                                                                                                  |
| | | +-----------------+             | |                                                                                                  |
| | +---------------------------------+ |                                                                                                  |
| | +--------------------------------+  |                                                                                                  |
| | | Trace                          |  |                                                                                                  |
| | +--------------------------------+  |                                                                                                  |
| | |  mesh   : /Mesh/object-1       |  |                                                                                                  |
| | |  nSpace : 1                    |  |                                                                                                  |
| | | +----------------------------+ |  |                                                                                                  |
| | | | Basis                      | |  |                                                                                                  |
| | | +----------------------------+ |  |                                                                                                  |
| | | |  is_continuous : 0         | |  |                                                                                                  |
| | | |  nComponents   : 1         | |  |                                                                                                  |
| | | |  nComponents1  : 1         | |  |                                                                                                  |
| | | |  nComponents2  : 1         | |  |                                                                                                  |
| | | |  nLocalDof     : 3         | |  |                                                                                                  |
| | | |  name          : lagrange  | |  |                                                                                                  |
| | | |  order         : 1         | |  |                                                                                                  |
| | | |  shape         : scalar    | |  |                                                                                                  |
| | | +----------------------------+ |  |                                                                                                  |
| | | +----------------+             |  |                                                                                                  |
| | | | Dof Table      |             |  |                                                                                                  |
| | | +----------------+             |  |                                                                                                  |
| | | |  nDof : 57630  |             |  |                                                                                                  |
| | | +----------------+             |  |                                                                                                  |
| | +--------------------------------+  |                                                                                                  |
| +-------------------------------------+                                                                                                  |
| +---------------------------------+                                                                                                      |
| | Time Discretization             |                                                                                                      |
| +---------------------------------+                                                                                                      |
| | +--------------+--------------+ |                                                                                                      |
| | | final time   | 2.500000e-01 | |                                                                                                      |
| | +--------------+--------------+ |                                                                                                      |
| | | initial time | 0.000000e+00 | |                                                                                                      |
| | +--------------+--------------+ |                                                                                                      |
| | | time step    | 2.000000e-02 | |                                                                                                      |
| | +--------------+--------------+ |                                                                                                      |
| | | type         | BDF          | |                                                                                                      |
| | +--------------+--------------+ |                                                                                                      |
| +---------------------------------+                                                                                                      |
+------------------------------------------------------------------------------------------------------------------------------------------+

============================================================
time simulation: 0.02s 
============================================================
 0 hdg.poisson KSP Residual norm 6.786386e+03
 1 hdg.poisson KSP Residual norm 3.494384e-10
 2 hdg.poisson KSP Residual norm 7.890788e-11
 3 hdg.poisson KSP Residual norm 5.713382e-11
============================================================
time simulation: 4.000000e-02s 
============================================================
 0 hdg.poisson KSP Residual norm 6.695973e+03
 1 hdg.poisson KSP Residual norm 3.392326e-10
 2 hdg.poisson KSP Residual norm 6.899868e-11
 3 hdg.poisson KSP Residual norm 4.974912e-11
============================================================
time simulation: 6.000000e-02s 
============================================================
 0 hdg.poisson KSP Residual norm 6.545492e+03
 1 hdg.poisson KSP Residual norm 3.236583e-10
 2 hdg.poisson KSP Residual norm 1.580683e-11
============================================================
time simulation: 8.000000e-02s 
============================================================
 0 hdg.poisson KSP Residual norm 6.337576e+03
 1 hdg.poisson KSP Residual norm 3.172950e-10
 2 hdg.poisson KSP Residual norm 3.731490e-11
============================================================
time simulation: 1.000000e-01s 
============================================================
 0 hdg.poisson KSP Residual norm 6.076527e+03
 1 hdg.poisson KSP Residual norm 3.020380e-10
 2 hdg.poisson KSP Residual norm 1.672100e-11
============================================================
time simulation: 1.200000e-01s 
============================================================
 0 hdg.poisson KSP Residual norm 5.768172e+03
 1 hdg.poisson KSP Residual norm 2.875859e-10
 2 hdg.poisson KSP Residual norm 2.726001e-11
============================================================
time simulation: 1.400000e-01s 
============================================================
 0 hdg.poisson KSP Residual norm 5.419050e+03
 1 hdg.poisson KSP Residual norm 2.840324e-10
 2 hdg.poisson KSP Residual norm 6.632669e-11
 3 hdg.poisson KSP Residual norm 4.809739e-11
============================================================
time simulation: 1.600000e-01s 
============================================================
 0 hdg.poisson KSP Residual norm 5.035070e+03
 1 hdg.poisson KSP Residual norm 2.554825e-10
 2 hdg.poisson KSP Residual norm 8.089654e-12
============================================================
time simulation: 1.800000e-01s 
============================================================
 0 hdg.poisson KSP Residual norm 4.622253e+03
 1 hdg.poisson KSP Residual norm 2.361476e-10
 2 hdg.poisson KSP Residual norm 1.848603e-11
============================================================
time simulation: 2.000000e-01s 
============================================================
 0 hdg.poisson KSP Residual norm 4.190434e+03
 1 hdg.poisson KSP Residual norm 2.129701e-10
 2 hdg.poisson KSP Residual norm 9.812686e-12
============================================================
time simulation: 2.200000e-01s 
============================================================
 0 hdg.poisson KSP Residual norm 3.754267e+03
 1 hdg.poisson KSP Residual norm 1.923188e-10
 2 hdg.poisson KSP Residual norm 2.638057e-11
============================================================
time simulation: 2.400000e-01s 
============================================================
 0 hdg.poisson KSP Residual norm 3.333780e+03
 1 hdg.poisson KSP Residual norm 1.727470e-10
 2 hdg.poisson KSP Residual norm 1.752302e-11
+-----------------------------------------------------------------------------------------------------------+
| Checkers : poisson                                                                                        |
+-----------------------------------------------------------------------------------------------------------+
| +-----------+-------------------------------+--------------+--------------+--------------+--------------+ |
| | check     | name                          | measure      | reference    | error        | tolerance    | |
| +===========+===============================+==============+==============+==============+==============+ |
| | [success] | Norm_potential_error_L2-error | 5.042606e-01 | 0.000000e+00 | 5.042606e-01 | 1.000000e+00 | |
| +-----------+-------------------------------+--------------+--------------+--------------+--------------+ |
| | [success] | Norm_flux_error_L2-error      | 9.119989e-01 | 0.000000e+00 | 9.119989e-01 | 1.000000e+00 | |
| +-----------+-------------------------------+--------------+--------------+--------------+--------------+ |
+-----------------------------------------------------------------------------------------------------------+
```

**Comment by prudhomm** [comment]
the next steps: 
 - get rid of fmilib which is a pain to maintain and put the fingers in.
 - add [feelpp/mo2fmu](https://github.com/feelpp/mo2fmu) support in feelpp to have fmu generation using dymola
 - pull_request

**Comment by prudhomm** [comment]
here are the results with om generated fmu using cvode solver

```
+-----------------------------------------------------------------------------------------------------------+
| Checkers : poisson                                                                                        |
+-----------------------------------------------------------------------------------------------------------+
| +-----------+-------------------------------+--------------+--------------+--------------+--------------+ |
| | check     | name                          | measure      | reference    | error        | tolerance    | |
| +===========+===============================+==============+==============+==============+==============+ |
| | [success] | Norm_potential_error_L2-error | 5.534465e-01 | 0.000000e+00 | 5.534465e-01 | 1.000000e+00 | |
| +-----------+-------------------------------+--------------+--------------+--------------+--------------+ |
| | [success] | Norm_flux_error_L2-error      | 8.261440e-01 | 0.000000e+00 | 8.261440e-01 | 1.000000e+00 | |
| +-----------+-------------------------------+--------------+--------------+--------------+--------------+ |
+-----------------------------------------------------------------------------------------------------------+
```

**Comment by prudhomm** [comment]
@romainhild @thomas-saigre  @mszopos here are the results of hdg coupledpoisson testcase with both dymola and openmodelica fmu tested with fmilib and fmi4cpp. fmi4cpp will replace fmilib 
here are the details computations which are all the same:
https://github.com/feelpp/feelpp/issues/1747#issuecomment-1006590230
interestingly the interface was slightly different between fmilib and fmi4cpp: to make a step we pass the current time for fmilib and the step size for fmi4cpp. Initially I passed the time to fmi4cpp and since the embedded solver in openmodelica is euler explicit as time increases I got a stability issues and massively wrong results, whereas with the dymola generated fmu the embedded solver was able  to cope with the large time steps. The wrong setup yielded results that were within error tolerance but the simulation time were completely different between the pde and ode simulators.
